ddev/src/ddev/cli/validate/metadata.py

```diff
@@ -0,0 +1,294 @@
+# (C) Datadog, Inc. 2023-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+from __future__ import annotations
+
+import csv
+import io
+import os
+from collections import defaultdict
+from typing import TYPE_CHECKING, DefaultDict, Set
+
+import click
+
+if TYPE_CHECKING:
+    from ddev.cli.application import Application
+
+
+def read_metadata_rows(metadata_file):
+    """
+    Iterate over the rows of a `metadata.csv` file.
+    """
+    with io.open(metadata_file, 'r', encoding='utf-8') as f:
+        reader = csv.DictReader(f, delimiter=',')
+
+        # Read header
+        reader._fieldnames = reader.fieldnames
+
+        for line_no, row in enumerate(reader, 2):
+            yield line_no, row
+
+
+def normalize_metric_name(metric_name):
+    """Copy pasted from the backend normalization code.
+    Extracted from dogweb/datalayer/metrics/query/metadata.py:normalize_metric_name
+    Metrics in metadata.csv need to be formatted this way otherwise, syncing metadata will fail.
+    Function not exported as a util, as this is different than AgentCheck.normalize. This function just makes sure
+    that whatever is in the metadata.csv is understandable by the backend.
+    """
+    from ddev.cli.validate import metadata_utils
+
+    if not isinstance(metric_name, str):
+        metric_name = str(metric_name)
+    metric_name = metadata_utils.METRIC_REPLACEMENT.sub("_", metric_name)
+    return metadata_utils.METRIC_DOTUNDERSCORE_CLEANUP.sub(".", metric_name).strip("_")
+
+
+def check_duplicate_values(current_check, line, row, header_name, duplicates, fail=None):
+    """Check if the given column value has been seen before.
+    Output a warning and return True if so.
+    """
+    if row[header_name] and row[header_name] not in duplicates:
+        duplicates.add(row[header_name])
+    elif row[header_name] != '':
+        message = f"{current_check}:{line} `{row[header_name]}` is a duplicate {header_name}"
+        if fail:
+            Application.display_error(message)
```

Comment:
I'm not sure if this case and the one below is ever tested because you must use an instance of the class Application otherwise this would produce an error

Reply:
fixed, added a test case to check for duplicate metric name and for when the optional flag is used